pkg/authentication/externalidp/awsiam/aws_client.go

```diff
@@ -0,0 +1,194 @@
+	return func(params *IAMAuthParams) {
+		params.RefreshInterval = refreshInterval
+	}
+}
+func WithURLPresignTTL(urlPresignTTL time.Duration) IAMAuthParamsOptions {
+	return func(params *IAMAuthParams) {
+		params.URLPresignTTL = urlPresignTTL
+	}
+}
+func WithTokenTTL(tokenTTL time.Duration) IAMAuthParamsOptions {
+	return func(params *IAMAuthParams) {
+		params.TokenTTL = tokenTTL
+	}
+}
+func WithTokenRequestHeaders(tokenRequestHeaders map[string]string) IAMAuthParamsOptions {
+	return func(params *IAMAuthParams) {
+		params.TokenRequestHeaders = tokenRequestHeaders
+	}
+}
+
+func NewIAMAuthParams(lakeFSHost string, opts ...IAMAuthParamsOptions) *IAMAuthParams {
+	headers := map[string]string{
+		"X-LakeFS-Server-ID": lakeFSHost,
+	}
+	p := &IAMAuthParams{
+		RefreshInterval:     DefaultRefreshInterval,
+		URLPresignTTL:       DefaultURLPresignTTL,
+		TokenTTL:            DefaultTokenTTL,
+		TokenRequestHeaders: headers,
+	}
+	for _, opt := range opts {
+		opt(p)
+	}
+	return p
+}
+
+func GenerateIdentityTokenInfo(ctx context.Context, params *IAMAuthParams, stsClient *sts.Client) (*AWSIdentityTokenInfo, error) {
+	url, err := PresignGetCallerIdentityFromAuthParams(ctx, params, stsClient)
+	if err != nil {
+		return nil, fmt.Errorf("generating sts presigned url: %w", err)
+	}
+	tokenInfo, err := ParsePresignedURL(url)
+	if err != nil {
+		return nil, fmt.Errorf("parsing credentials sts presigned url: %w", err)
+	}
+	return tokenInfo, nil
+}
+
+func ParsePresignedURL(presignedURL string) (*AWSIdentityTokenInfo, error) {
+	parsedURL, err := url.Parse(presignedURL)
+	if err != nil {
+		return nil, err
+	}
+
+	queryParams := parsedURL.Query()
+	credentials := queryParams.Get(AuthCredentialKey)
+	splitedCreds := strings.Split(credentials, "/")
+	if len(splitedCreds) < minLengthSplitCreds {
+		return nil, fmt.Errorf("min length for query param not '%d' ('%s'): %w", minLengthSplitCreds, splitedCreds, ErrInvalidCredentialsFormat)
+	}
+	calculatedRegion := splitedCreds[2]
+	accessKeyID := splitedCreds[0]
+	return &AWSIdentityTokenInfo{
+		Method:             "POST",
+		Host:               parsedURL.Host,
+		Region:             calculatedRegion,
+		Action:             AuthAction,
+		Date:               queryParams.Get(AuthDateKey),
+		ExpirationDuration: queryParams.Get(AuthExpiresKey),
+		AccessKeyID:        accessKeyID,
+		Signature:          queryParams.Get(AuthSignatureKey),
+		SignedHeaders:      strings.Split(queryParams.Get(AuthSignedHeadersKey), ";"),
+		Version:            queryParams.Get(AuthVersionKey),
+		Algorithm:          queryParams.Get(AuthAlgorithmKey),
+		SecurityToken:      queryParams.Get(AuthSecurityTokenKey),
+	}, nil
+}
+
+func PresignGetCallerIdentityFromAuthParams(ctx context.Context, params *IAMAuthParams, stsClient *sts.Client) (string, error) {
+	setHTTPHeaders := func(requestHeaders map[string]string, ttl time.Duration) func(*middleware.Stack) error {
+		middlewareName := "AddHeaders" + uuid.New().String()
```

Comment:
This is not a bug but a potential for one if the presignClient moves somewhere else and will be injected from the outside.
Should make the name constant, it'll make sure only single middleware of this type will be added to the presign client otherwise will error.